File: Unit2-10/Typescript/Vehicle.ts
/**
*
* This class creates a vehicle
*
* By:      Evan Beaudoin
* Version: 1.2
* Since:   2024-03-28
*/

export abstract class Vehicle {

    // Creates the base vehicle
    constructor(
      // Class variables
        protected _colour: string, protected _maxSpeed: number, protected _wheels: number, protected _speed: number
    ) {}

    // Gets the colour of the vehicle
    get colour(): string {
        return this._colour
    }

    // Gets the current speed of the vehicle
    get speed(): number {
        return this._speed
    }

    // Gets the maximum speed of the vehicle
    get maxSpeed(): number {
        return this._maxSpeed
    }

    // Gets the number of wheels on the vehicle
    get wheels(): number {
        return this._wheels
    }

    // Creates a string about the status of the car
    abstract get status(): string

    // Sets the colour of the vehicle
    set colour(newColour: string) {
        this._colour = newColour
    }

    // Accelerates the car, changing the speed
    abstract accelerate(accelPower: number, accelTime: number): void

    // Brakes the car, changing the speed
    abstract brake(brakePower: number, brakeTime: number, airPressure: number): void
}

File: Unit2-10/Java/Vehicle.java
/*
* This class creates a vehicle
*
* @author  Evan Beaudoin
* @version 1.2
* @since   2024-03-28
*/

/**
 * This is the class.
*/
abstract class AbstractVehicle {
    /**
     * Variable for the class.
     *
     * @param colour of the vehicle
    */
    private String colour = "";

    /**
     * Variables for the class.
     *
     * @param speed of the vehicle
    */
    private double speed;

    /**
     * Variables for the class.
     *
     * @param maxSpeed of the vehicle
    */
    private double maxSpeed;

    /**
     * Variables for the class.
     *
     * @param wheels on the vehicle
    */
    private int wheels;

    /**
     * Creates the base vehicle.
     *
     * @param startColour the base colour
     * @param startMaxSpeed the maximum speed of the vehicle
     * @param startWheels the number of wheels on the vehicle
    */
    AbstractVehicle(String startColour, double startMaxSpeed, int startWheels) {
        this.colour = startColour;
        this.speed = 0;
        this.maxSpeed = startMaxSpeed;
        this.wheels = startWheels;
    }

    /**
     * Gets the colour of the vehicle.
     *
     * @return the current colour
    */
    String getColour() {
        return this.colour;
    }

    /**
     * Gets the speed of the vehicle.
     *
     * @return the current speed
    */
    double getSpeed() {
        return this.speed;
    }

    /**
     * Gets the maximum speed of the vehicle.
     *
     * @return the maximum speed
    */
    double getMaxSpeed() {
        return this.maxSpeed;
    }

    /**
     * Gets the number of wheels on the vehicle.
     *
     * @return the total number of wheels
    */
    int getWheels() {
        return this.wheels;
    }

    /**
     * Creates a string about the status of the vehicle.
     *
     * @return the string about the vehicle's status
    */
    abstract String getStatus();

    /**
     * Sets the colour of the vehicle.
     *
     * @param newColour of the vehicle
    */
    void setColour(String newColour) {
        this.colour = newColour;
    }

    /**
     * Sets the speed of the vehicle.
     *
     * @param newSpeed of the vehicle
    */
    void setSpeed(double newSpeed) {
        this.speed = newSpeed;
    }

    /**
     * Accelerates the vehicle, which changes the speed.
     *
     * @param accelPower power of the acceleration
     * @param accelTime time of the acceleration
    */
    void accelerate(double accelPower, double accelTime) {
        this.speed = (accelPower * accelTime) + this.speed;
        if (this.speed > this.maxSpeed) {
            this.speed = this.maxSpeed;
        }
    }

    /**
     * Brakes the vehicle, which changes speed.
     *
     * @param brakePower power of the brake
     * @param brakeTime time of the brake
    */
    void brake(double brakePower, double brakeTime) {
        this.speed = this.speed - (brakePower * brakeTime);
        if (this.speed < 0) {
            this.speed = 0;
        }
    }
}